Madness/Parser.swift

```diff
@@ -5,38 +5,41 @@
 /// \param Tree  The type of parse tree generated by the parser.
 public enum Parser<C: CollectionType, Tree> {
 	/// The type of parser combinators.
-	public typealias Function = (C, C.Index) -> Result
+	public typealias Function = (C, SourcePos<C.Index>) -> Result
 
 	/// The type produced by parser combinators.
-	public typealias Result = Either<Error<C.Index>, (Tree, C.Index)>
+	public typealias Result = Either<Error<C.Index>, (Tree, SourcePos<C.Index>)>
 }
 
 /// Parses `input` with `parser`, returning the parse trees or `nil` if nothing could be parsed, or if parsing did not consume the entire input.
 public func parse<C: CollectionType, Tree>(parser: Parser<C, Tree>.Function, input: C) -> Either<Error<C.Index>, Tree> {
-	return parser(input, input.startIndex).flatMap { $1 == input.endIndex ? .right($0) : .left(.leaf("finished parsing before end of input", $1)) }
+	let result = parser(input, SourcePos(index: input.startIndex))
+	
+	return result.flatMap { tree, sourcePos in
+		return sourcePos.index == input.endIndex
+			? .Right(tree)
+			: .Left(.leaf("finished parsing before end of input", sourcePos))
+	}
 }
 
 public func parse<Tree>(parser: Parser<String.CharacterView, Tree>.Function, input: String) -> Either<Error<String.Index>, Tree> {
-	return parser(input.characters, input.startIndex).flatMap { $1 == input.endIndex ? .right($0) : .left(.leaf("finished parsing before end of input", $1)) }
+	return parse(parser, input: input.characters)
 }
 
 // MARK: - Terminals
 
 /// Returns a parser which never parses its input.
 public func none<C: CollectionType, Tree>(string: String = "no way forward") -> Parser<C, Tree>.Function {
-	return { _, index in Either.left(Error.leaf(string, index)) }
+	return { _, sourcePos in Either.left(Error.leaf(string, sourcePos)) }
 }
 
-/// Returns a parser which parses any single character.
-public func any<C: CollectionType>(input: C, index: C.Index) -> Parser<C, C.Generator.Element>.Result {
-	if index != input.endIndex {
-		let parsed = input[index]
-		let next = index.successor()
-		
-		return .Right((parsed, next))
-	} else {
-		return .Left(Error.leaf("", index))
-	}
+// Returns a parser which parses any single character.
+public func any<C: CollectionType>(input: C, sourcePos: SourcePos<C.Index>) -> Parser<C, C.Generator.Element>.Result {
+	return satisfy { c in true }(input, sourcePos)
+}
+
+public func any(input: String.CharacterView, sourcePos: SourcePos<String.Index>) -> Parser<String.CharacterView, Character>.Result {
+	return satisfy { c in true }(input, sourcePos)
```

Comment:
On second thought, we need this overload so any increments newline/columns when parsing `CharacterViews`.